### DISCUSSION TITLE
moving non-conforming meshes, and/or overset capabilities

### DISCUSSION THREAD

**Discussion by slopezcastano** [idaholab/moose#24154] [Q&A General] [answered]
Hi everyone,

  I was just wondering, as a possible newcomer and user of MOOSE with some C++/CFD experience, whether is possible to handle  moving non-conformal meshes (via interfaces, think of a moving propeller attached to a boat), and/or overset within the framework. I'm mostly interested in fluid flow simulations.   

Regards, 

Santiago

**Comment by GiudGiud** [comment]
Hello

While we do have fluid flow capabilities, non-conformal meshes are only handled through custom interface treatments in MOOSE and we have not done much of that for fluid flow.

Were you looking for finite element or finite volume for fluid flow?

Guillaume

**Reply by slopezcastano** [reply]
Hi Guillaume,

In this particular case, my interest in non-conformal sliding meshes is mostly for fluid flows using FVM. I am also be interested on coupled solvers (as opposed to segregated) in FVM/FEM (for the study of polydisperse multiphase flows); this is one of the strongest reasons regarding my interest in MOOSE (SNES PETSc, under the hood but more abstract interface). I mean, there are some FVM libraries on the wild, but have a rather incipient treatment of coupled solvers and don't support PETSc. 

Basically, I'd like to start with simulations using either sliding meshes (non-conformal, with non trivial topologies) or chimera (overset), if possible using INSE. I, generally, don't shy away from having to develop some code if needed but I'd like to get an idea of how much "time" it will require for me to get to the point I'm interested to be.

Many thanks for the attention,

Santiago

**Reply by GiudGiud** [reply]
So we provide:
- several FE & FV discretizations of the Navier Stokes equations for various flow regimes
- a few turbulence models, with k-epsilon soon (coming months) to be merged for finite volume
- examples of efficient preconditioning of flow & multiphysics equations through our interface to PETSc
- limited multiphase capabilities for gas-liquid (in development) and liquid-solid
- coupled solver is the default for now, we're adding a segregated solver in the coming months
Summary of flow capabilities: https://mooseframework.inl.gov/modules/navier_stokes/index.html
- the displaced mesh capabilities are shown best on our gallery
https://mooseframework.inl.gov/gallery.html

I'll let @lindsayad comment on a time estimate for dev of sliding meshes with FE flow

**Reply by lindsayad** [reply]
I would have to read about overset; I’m not familiar with it. We do have mortar methods which we use for sliding frictional contact in solid mechanics simulations. We’ve applied mortar to finite volume discretizations for gap heat transfer. Mortar has been our recent go-to for non conforming meshes. In the past we’ve also used node to face constraints (only in finite element contexts).

**Reply by lindsayad** [reply]
Looking briefly at overset I think a multi app approach with projections of fields between the different meshes would be the easiest approach with MOOSE

**Reply by lindsayad** [reply]
But that wouldn’t be in line with your principal interest of a fully coupled solve. At the libMesh implementation level we have one EquationSystems object per Mesh(Base). It’s an interesting question of how we might do this with two disconnected meshes in a single MeshBase object with projections between the two. @roystgnr have you ever had anyone interested in this type of thing? Thinking about this made me think of IBAMR but I think these are fundamentally different methods

**Reply by lindsayad** [reply]
Both mortar and IBM are surface couplings whereas you want to do volume couplings…

**Reply by slopezcastano** [reply]
Hi @lindsayad,

   Mortar couplings are a good start! For the first part of the study I have in mind "sliding meshes" are enough (noise propagation on a confined marine propeller). For this I don't need "multi-mesh equation coupling" or overset, since I'll go for classical INSE-LES and the topology of the disjoint meshes can be made non-overlapping.

  ... But, down the road, I'd like to study cavitation on said propeller (polydisperse multiphase models) which is more stable when the equations are coupled (Conservation of mass becomes critical between phases, hence a segregated approach tends to "break down" when the dispersed n-phase approaches zero). Notice here that level-set may be used for the resolution of the bubbles.

Finally I'd like to start adding more "topological features" (rudders, propellers) which are obviously "simpler" to mesh as disjointed, separate volume meshes.

At the risk of taking too much of your time, I'd like to expand a bit on the discussion we have now:

   1. **Mortar Methods**: If I understood correctly you guys have a CHT case from which I might be able to use as "template" for a multi-region INSE solver.  Is that correct?
   2. **Level-Set**: As I see in the examples, there is a PoC for level-set, however there are some suggestions on how to "generalize" it. is there any attempt now to build a two-phase (or multiphase) solver within DG or FVM?
   3. **Volume-coupling (overlapping meshes, Schwartz method, Overset) and Equation coupling**: Is it necessary within MOOSE to also couple the "interpolation contribution" onto the "stiffness matrix", or can it be possible to handle the "inter-mesh" coupling explicitely? I understand there will be an interpolation and first-order-in-time error produced there but I suspect adding the interpolation stencils onto the matrix may be quite time consuming (one might need to search the nodes/centroids both in the domain as well as in the matrix) depending how the matrix is also stored in memory (I talk as a complete newbie with MOOSE, but as an avg user of petsc in FVM). 

Anyway, I really thank you for taking the time to answer my questions (not very common in some FOSS CFD forums). I'll download the code and start trying it out so to get an idea on the workflow there.

Kind regards,

Santiago

**Reply by roystgnr** [reply]
Paul Bauman put some code in GRINS (another libMesh-based physics code) using overset grids for solid-fluid interactions.  There was no need to put two separate meshes in the same `MeshBase` there, though; the solid and fluid domains were just two overlapping subdomains of the same `Mesh`.  A few libMesh features (like feeding subdomain id sets to a `PointLocator`) are designed with that in mind, and IIRC Paul added some lookup caching on top of that (the `PointLocatorTree` is O(log N) and I guess switching to O(1) makes a difference when you're looking at every quadrature point...).  He also wrote a custom `GhostingFunctor` subclass to use as a coupling functor to handle the additional sparsity entries (and thereby incidentally handle the algebraic and geometric ghosting; man I'm proud of that API...).  I don't recall what he was doing for quadrature on the mismatching meshes, though.

**Reply by roystgnr** [reply]
We could ping Paul if we have any questions critical enough.  But GRINS is no longer officially maintained, and yet he still gives me a hand keeping it up to date when I run into any issues, and I still rely on it every now and then for added test coverage, so I don't want to abuse that privilege too hard.  His code is all LGPL, so we've got the goahead if there's anything like that ghosting functor that we want to bring into libMesh or MOOSE.

**Reply by lindsayad** [reply, marked as answer]
> 1. **Mortar Methods**: If I understood correctly you guys have a CHT case from which I might be able to use as "template" for a multi-region INSE solver.  Is that correct?

Yes, we have conjugate heat transfer, mechanical contact, etc. But only heat transfer demonstrated in a finite volume context, so that is probably the best template.

> * is there any attempt now to build a two-phase (or multiphase) solver within DG or FVM?

We have two-phase flow in a 7 equation model using rDG(P0P1) (so finite volume) in the relap-7 application, but we'd like to get something in our MOOSE modules. You can see https://github.com/idaholab/moose/issues/22637 for discussion on the topic. In short, our team at Idaho National Laboratory is not currently doing anything, but we'd love to see movement on the topic.

> 3\. olume-coupling (overlapping meshes, Schwartz method, Overset) and Equation coupling

Based on @roystgnr's response, I think we could definitely so something here
